SAMPLEs = ["Ctrl", "CoV"]
REPs = ["rep1", "rep2"]

rRNA_INDX = "data/index/STAR/rRNA"
pre_rRNA_INDX = "/home/wangdehe/common/rRNA/star_indx/human_pre"
chkSab_INDX = "data/index/STAR/chkSab"
WIV04_INDX = "data/index/STAR/WIV04"
chkSab_WIV04_INDX = "data/index/STAR/chkSab_WIV04"
hg38_INDX = "data/index/STAR/hg38"

GENOME_FA = "data/genome/chkSab.fa"
GENOME_2bit = "data/genome/chkSab.2bit"
CHROM_SIZE = "data/genome/chrom.sizes"
REF_BED = "data/genome/chkSab.bed12"
REF_GTF = "data/genome/chkSab.gtf"
REF_GFF = "data/genome/chkSab.gff3"
REF_GENE_BED = "data/genome/chkSab.gene.bed6"
REF_GENE_INFO = "data/genome/chkSab.info.tsv"
REF_INTRON = "data/genome/chkSab.intron.bed6"
hg38_REF_GTF = "data/genome/hg38_gencode.v32.gtf"
UCSC_SARS2_CHROM_SIZE = "data/genome/wuhCor1.sizes"

RCS_FA = "data/genome/RCS_RENAMED.fasta"

WIV04_GENOME_FA = "data/genome/WIV04.fa"
WIV04_GENE_BED = "data/genome/WIV04.bed"
WIV04_CHROM_SIZE = "data/genome/WIV04.sizes"
WIV04_GENE_GTF = "data/genome/WIV04.gtf"

MERGE_GENOME_FA = "data/genome/chkSab_WIV04.merge.fa"
MERGE_SIZE = "data/genome/chkSab_WIV04.merge.sizes"

